rir/src/utils/FunctionWriter.h

```diff
@@ -5,80 +5,69 @@
 #include "interpreter/runtime.h"
 #include "ir/BC_inc.h"
 
+#include "R/Preserve.h"
 #include "ir/CodeVerifier.h"
 #include "runtime/Function.h"
+#include "utils/Pool.h"
 
 #include <iostream>
+#include <vector>
 
 namespace rir {
 
 class FunctionWriter {
   public:
     typedef unsigned PcOffset;
-    constexpr static unsigned initialSize = 2 * sizeof(Function);
+
+    size_t functionSize; // size of the Function, including pointers to its code
+                         // objects
 
     Function* function;
 
-    size_t capacity;
+    std::vector<SEXP> codeVec;
+
+    Preserve preserve;
+
+    FunctionWriter() : functionSize(sizeof(Function)), function(nullptr) {}
 
-    FunctionWriter() = delete;
-    // also copy and move ctors = default ?
+    ~FunctionWriter() {}
 
-    static FunctionWriter create() {
-        assert(initialSize > sizeof(Function));
-        assert(initialSize % sizeof(int) == 0);
-        SEXP store = Rf_allocVector(EXTERNALSXP, initialSize);
+    void finalize() {
+        assert(function == nullptr && "Trying to finalize a second time");
+
+        size_t dataSize = codeVec.size() * sizeof(SEXP);
+        assert(functionSize == sizeof(Function) + dataSize);
+        assert(functionSize % sizeof(int) == 0);
+
+        SEXP store = Rf_allocVector(EXTERNALSXP, functionSize);
         void* payload = INTEGER(store);
+        Function* fun = new (payload) Function(functionSize, codeVec);
+        preserve(fun->container());
```

Comment:
The `Function` constructor doesn't trigger gc, but wouldn't it be safer to `preserve(store)` immediately after allocating it?

Reply:
Turns out that's wrong. We can only preserve after initializing the object.